Functionalize WS parsing code

gleber opened the pull request:
Hello Roberto

Here is functionalized version of WS parsing code, which allows to use WS parsing code outside of misultin (and allows to avoid an additional process).

Best,
Gleb

gleber committed:
refactor required WS headers into separate functions

gleber committed:
export required_headers/0 of each websocket version

gleber committed:
rewrite websocket management to use foldl-like functional style, part 2 - hixie-68

gleber committed:
rewrite websocket management to use foldl-like functional style, part 3 - hybi-10_17

gleber committed:
rewrite websocket management to use foldl-like functional style, part 1

ostinelli commented:
hello gleber,

thank you for your submit. though, i'm not sure i want to do this.

the websocket protocol has been evolving quite a lot, and since last month it [finally made it to RFC](http://www.rfc-editor.org/authors/rfc6455.txt). it has been a real pain to maintain all the different evolutions, and even having the current situation in misultin (behaviours) for websockets is something i want to get rid of asap (even though this is obviously going to take some time, waiting for all the major browsers to catch up). the refactored code you are proposing adds to the overhead and complications, which is something that i don't want to do.

afa the 'avoid an additional process' part, i don't quite follow: the websocket module in your refactoring [is still spawning another process](https://github.com/gleber/misultin/blob/functionalize-ws-code/src/misultin_websocket.erl#L70).

i like your suggestion though, and i plan to look into single process for websockets this weekend, as you also suggested in another thread. so we'll probably be able to discuss this again if you feel you'd like to :)

cheers,

r.

gleber commented:
In previous email I described, that I will not rewrite whole misultin, instead I just rewritten only WS parsing modules, and changed behaviour specification accordingly (in fact only handle_data/5 function). As you can see inside implementations of misultin_websocket behaviour - I got rid of WsHandleLoopPid there. Instead [a callback](https://github.com/gleber/misultin/blob/functionalize-ws-code/src/misultin_websocket.erl#L175) is passed [throughout the code](https://github.com/gleber/misultin/blob/functionalize-ws-code/src/misultin_websocket_draft-hixie-68.erl#L118):

Also this callback works similarly as in foldl function. I.e. it's possible to pass around "mutable" state or/and accumulated received messages too. Here's how I use it in our code: https://gist.github.com/1444187 (this is distilled, but should show the idea). 

I think this is a pretty good start for rewriting misultin to use only one process for websocket connections, since parsing modules are no longer dependent on additional process, so no changed would be necessary there - you just pass another Fun there. And since it's foldl-like, it would be enough for all possible functional abstractions over parsing a stream of bytes.

Also I don't really know any other way (which wouldn't be equivalent to what I did) to rewrite WS parsing code NOT to use second process. I believe that your way to rewrite to use just single process would end up with very conceptually similar code as I have posted in a gist and in my branch :) Is there any other way?

Best,
Gleb

ostinelli commented:
as said, i'll consider using a single process for websocket handling, which is an idea i like and that i've already started digging into [there's a branch for that].

this commit, however, assumes things that don't exist yet, and i have no reason in bringing this or some equivalent solution in for the time being. i'll close this issue for now, and will reopen it if necessary.

thank you,

r.
